R/MsExperiment.R

```diff
@@ -214,48 +214,99 @@ setMethod("readMsObject", signature(object = "MsExperiment",
+              s_files <- all_fls[grepl("^s_", all_fls)]
+              sample_info <- read.table(paste0(pth, s_files),
+                                        header = TRUE, sep = "\t",
+                                        check.names = FALSE)
+
+              # merging
+              ord <- match(assay_data$`Sample Name`, sample_info$`Sample Name`)
+              merged_data <- cbind(assay_data, sample_info[ord, ])
+              names(merged_data) <- gsub(" ", "_", names(merged_data))
+              if (keepProtocol || keepOntology || simplify)
+                  merged_data <- .clean_merged(x = merged_data,
+                                               keepProtocol = keepProtocol,
+                                               keepOntology = keepOntology,
+                                               simplify = simplify)
+
+              ## Assemble object
+              object@spectra <- Spectra::Spectra(mtblsId = param@mtblsId,
+                                                 source = MsBackendMetaboLights::MsBackendMetaboLights(),
+                                                 assayName = selected_assay,
+                                                 filePattern = param@filePattern)
+
+              ## sample to spectra link
+              fl <- object@spectra@backend@spectraData[1, "derived_spectral_data_file"]
+              nme <- colnames(merged_data)[which(merged_data[1, ] == fl)]
+              merged_data <- merged_data[grepl(param@filePattern,
+                                               merged_data[, nme]), ]
 
               object@sampleData <- DataFrame(merged_data)
+              object <- MsExperiment::linkSampleData(object,
+                                                     with = paste0("sampleData.",
+                                                                nme,
+                                                                "= spectra.derived_spectral_data_file"))
               validObject(object)
               object
           })
+
+
+#####HELPERS
+
+#' function that takes the extra parameters and clean the metadata if asked by
+#' the user.
+#'
+#' Note:  the subsetting of the merged data is done here, which WILL rename the
+#' duplicated columns. I could fix that by first transforming the data.frame into
+#' a list but I am not sure that it is useful.. The user might do some
+#' subsetting too later and then the same thing will happen. Might as well have
+#' it from the beginning.
+#'
+#' Note2: I would move that function later, keeping it her for review to help
+#' clarity
+#'
+#' @noRd
+.clean_merged <- function(x, keepProtocol, keepOntology, simplify) {
+    # remove ontology
+    if (!keepOntology)
+        x <- x[, -which(grepl("Term", names(x)))]
+
+    # remove protocol
+    if (!keepProtocol)
+        x <- x[, -which(grepl("Protocol|Parameter", names(x)))]
+
+    # remove duplicated columns contents and NAs
+    if (simplify) {
+        x <- x[, !duplicated(as.list(x))]
```

Comment:
A looot of columns (with different colnames) had the sample names in them so this is to remove them. It also remove columns full of NAs actually but it leaves one, which mean i have to still do the line below, it's not super elegant in my opinion but  could not think of anything else.